ext/intl/grapheme/grapheme_string.c

```diff
@@ -375,7 +375,9 @@ PHP_FUNCTION(grapheme_substr)
 		RETURN_THROWS();
 	}
 
-	if ( OUTSIDE_STRING(lstart, str_len)) {
+	if (str_len == 0 && lstart == 0) {
+		RETURN_EMPTY_STRING();
+	} else if (OUTSIDE_STRING(lstart, str_len)) {
```

Comment:
I think the real problem is in OUTSIDE_STRING, which should only consider lstart > str_len to be out-of-bounds, not lstart >= str_len. "One past the end" is generally always considered a valid offset (also in strpos etc).

Though this is just a "definitely wrong" check, as this is a grapheme cluster offset, the actual offset check happens later...

Reply:
I think even lstart > str_len should be allowed for grapheme_substr() (and probably other functions). Whether it should return "" or false in that case is up for debate but throwing an exception IMHO cripples the API.

Reply:
I'll look into this tonight

Reply:
I've applied a fix for the main problem in https://github.com/php/php-src/commit/1312c417dea0e86e8af7f98bc24b3dd055e51e14 (there are still issues for the non-ASCII case). For the grapheme_substr() case, I would first like to decide on a final behavior for substr(), before we adjust this one (#6182).